src/librustc/infer/error_reporting/mod.rs

```diff
@@ -704,7 +707,31 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                         &format!(" ({})", values.expected.sort_string(self.tcx)),
                         &format!(" ({})", values.found.sort_string(self.tcx)));
                 }
-                (_, false,  _) => {
+                (_, false, _) => {
+                    if let Some(exp_found) = exp_found {
+                        let (def_id, ret_ty) = match exp_found.found.sty {
+                            TypeVariants::TyFnDef(def, _) => {
+                                (Some(def), Some(self.tcx.fn_sig(def).output()))
+                            }
+                            _ => (None, None)
+                        };
+
+                        let exp_is_struct = match exp_found.expected.sty {
+                            TypeVariants::TyAdt(def, _) => def.is_struct(),
+                            _ => false
+                        };
+
+                        if let (Some(def_id), Some(ret_ty)) = (def_id, ret_ty) {
+                            if exp_is_struct && exp_found.expected == ret_ty.0 {
+                                let message = format!(
+                                    "did you mean `{}(/* fields */)`?",
+                                    self.tcx.item_path_str(def_id)
+                                );
+                                diag.help(&message);
```

Comment:
Last thing and this will be ready for merging (r=me), could you change `diag.help(&message)` here for `span_label(cause.span, &message)`?

That way the output will be:

```
error[E0308]: mismatched types
  --> $DIR/issue-35241.rs:13:20
   |
13 | fn test() -> Foo { Foo }
   |              ---   ^^^ 
   |              |     |
   |              |     expected struct `Foo`, found fn item
   |              |     did you mean `Foo(/* fields */)`?
   |              |
   |              expected `Foo` because of return type
   |
   = note: expected type `Foo`
              found type `fn(u32) -> Foo {Foo::{{constructor}}}`
```

---

Also, while you're at it, could you squash your PR into a single commit?